libnmstate/nm/active_connection.py

```diff
@@ -20,162 +20,369 @@
+            self._nm_dev,
+            specific_object,
+            self._ctx.cancellable,
+            self._activate_profile_callback,
+            user_data,
+        )
+
+    @staticmethod
+    def wait(ctx, nm_ac, nm_dev):
+        obj = ProfileActivation(
+            ctx,
+            nm_dev.get_iface(),
+            get_iface_type(nm_dev),
+            None,
+            nm_dev,
+        )
+        obj._nm_ac = nm_ac
+        obj._action = (
+            f"Waiting activation of {obj._iface_name} {obj._iface_type}"
+        )
+        ctx.register_async(obj._action)
+        obj._wait_profile_activation()
+
+    def _activate_profile_callback(self, nm_client, result, _user_data):
+        nm_ac = None
+        if self._ctx.is_cancelled():
+            self._activation_clean_up()
             return
+        try:
+            nm_ac = nm_client.activate_connection_finish(result)
+        except Exception as e:
+            self._ctx.fail(
+                NmstateLibnmError(f"{self._action} failed: error={e}")
+            )
+
+        if nm_ac is None:
+            self._ctx.fail(
+                NmstateLibnmError(
+                    f"{self._action} failed: "
+                    "error='None return from activate_connection_finish()'"
+                )
+            )
+        else:
+            logging.debug(
+                f"Connection activation initiated: iface={self._iface_name} "
+                f"type={self._iface_type} con-state={nm_ac.get_state()}"
+            )
+            self._nm_ac = nm_ac
+            self._nm_dev = get_nm_dev(
+                self._ctx, self._iface_name, self._iface_type
+            )
+            self._wait_profile_activation()
 
-        if self._act_con != act_connection:
-            raise NmstateLibnmError(
-                "When deactivating active connection, the newly get "
-                f"NM.ActiveConnection {act_connection}"
-                f"is different from original request: {self._act_con}"
+    def _wait_profile_activation(self):
+        if is_activated(self._nm_ac, self._nm_dev):
+            logging.debug(
+                "Connection activation succeeded: "
+                f"iface={self._iface_name}, type={self._iface_type}, "
+                f"con_state={self._nm_ac.get_state()}, "
+                f"dev_state={self._nm_dev.get_state()}, "
+                f"state_flags={self._nm_ac.get_state_flags()}"
+            )
+            self._activation_clean_up()
+            self._ctx.finish_async(self._action)
+        elif is_activating(self._nm_ac, self._nm_dev):
+            if self._nm_ac:
+                self._wait_nm_ac_activation()
+            if self._nm_dev:
+                self._wait_nm_dev_activation()
+            if not self._nm_ac and not self._nm_dev:
+                self._ctx.fail(
+                    NmstateInternalError(
+                        f"{self._action} failed: no nm_ac or nm_dev"
+                    )
+                )
```

Comment:
I think this is missing `self._activation_clean_up` before failing, maybe there could be a condition race when rolling back. Not sure about it.

Reply:
There is no `nm_ac` or `nm_dev`, nothing need to be clean up.